README.md
[![Yii2](https://img.shields.io/badge/required-Yii2_v2.0.40-blue.svg)](https://packagist.org/packages/yiisoft/yii2)
[![Downloads](https://img.shields.io/packagist/dt/wdmg/yii2-admin.svg)](https://packagist.org/packages/wdmg/yii2-admin)
[![Packagist Version](https://img.shields.io/packagist/v/wdmg/yii2-admin.svg)](https://packagist.org/packages/wdmg/yii2-admin)
![Progress](https://img.shields.io/badge/progress-ready_to_use-green.svg)
[![GitHub license](https://img.shields.io/github/license/wdmg/yii2-admin.svg)](https://github.com/wdmg/yii2-admin/blob/master/LICENSE)

<img src="./docs/images/yii2-admin.png" width="100%" alt="Administrative panel for Butterfly.CMS" />

# Yii2 Admin Module
Administrative panel for [Butterfly.CMS](https://github.com/wdmg/butterfly.cms).
                                                                                                     
This module is an integral part of the [Butterfly.СMS](https://butterflycms.com/) content management system, but can also be used as an standalone extension.

Copyrights (c) 2019-2023 [W.D.M.Group, Ukraine](https://wdmg.com.ua/)

# Requirements 
* PHP 5.6 or higher
* Yii2 v.2.0.40 and newest
* [Yii2 Base](https://github.com/wdmg/yii2-base) module (required)
* [Yii2 Users](https://github.com/wdmg/yii2-users) module (required)

# Installation
To install the module, run the following command in the console:

`$ composer require "wdmg/yii2-admin"`

After configure db connection, run the following command in the console:

`$ php yii admin/init`

And select the operation you want to perform:
  1) Apply all modules migrations
  2) Revert all modules migrations

# Migrations
In any case, you can execute the migration and create the initial data, run the following command in the console:

`$ php yii migrate --migrationPath=@vendor/wdmg/yii2-admin/migrations`

# Configure

To add a module to the project, add the following data in your configuration file:

    'modules' => [
        ...
        'admin' => [
            'class' => 'wdmg\admin\Module',
            'routePrefix' => 'admin',
            'checkForUpdates' => true, // boolean, the flag if updates check turn on
            'cacheExpire' => 3600, // integer, the time to expire cache
            'multiSignIn' => false, // not allow by default
            'sessionTimeout' => 900, // 15 min.
            'customLocales' => [ // expanding the list of language locales for searching translations
                'uk-UA' => 'Українська',
            ],
            'customSupportModules' => [ // expanding the list of modules available for installation and download
                'wdmg/yii2-example',
            ],
            'customSidebarMenu' => [ // extending the sidebar menu list
                [
                    'label' => 'Example',
                    'icon' => 'fa fa-fw fa-bars',
                    'url' => ['/admin/example/default'],
                    'order' => 10,
                ]
            ],
            'customCreateMenu' => [ // expanding the creation menu list
                'wdmg/yii2-example' => [
                    'label' => 'Add new example',
                    'url' => ['/admin/example/default/create']
                ]
            ],
            'useSphinxSearch' => true, // boolean, the flag for configuration Sphinx Search
            'sphinxSearchConf' => [ // configuration of Sphinx Search daemon
                'dsn' => "mysql",
                'host' => "127.0.0.1",
                'port' => "9306",
                'username' => "",
                'password' => "",
            ]
        ],
        ...
    ],

# Routing

| Link to route (without prettyurl)                     | Link to route (prettyurl enabled)     | Description                          |
|:----------------------------------------------------- |:------------------------------------- |:------------------------------------ |
| http://example.com/index.php?r=admin                  | http://example.com/admin              | Main dashboard                       |
| http://example.com/index.php?r=admin/admin/login      | http://example.com/admin/login        | Auth to dashboard                    |
| http://example.com/index.php?r=admin/admin/restore    | http://example.com/admin/restore      | Restore access password              |

Use the `Module::dashboardNavItems()` method of the module to generate a navigation items list for NavBar, like this:

    <?php
        echo Nav::widget([
        'options' => ['class' => 'navbar-nav navbar-right'],
            'label' => 'Modules',
            'items' => [
                Yii::$app->getModule('admin')->dashboardNavItems(),
                ...
            ]
        ]);
    ?>

# Status and version [ready to use]
* v.1.4.6 - Fix sidebar, terminal link
* v.1.4.5 - Count fixed in sidebar menu
* v.1.4.4 - Add task and tickets counters

CHANGELOG.md
Changelog
=========

## 1.4.6 (2023-07-06)
 * Fix sidebar, terminal link

## 1.4.5 (2023-06-28)
 * Count fixed in sidebar menu

## 1.4.4 (2023-06-25)
 * Add task and tickets counters

## 1.4.3 (2023-06-25)
 * Add server date and time to dashboard UI

## 1.4.2 (2023-06-21)
 * Fix add to favourites

## 1.4.1 (2023-06-21)
 * Fix Users::getOptions for not auth
 * Dashboard favourites list
 * Update node sass
 * Fix lang, update dependencies

## 1.4.0 (2023-03-30)
 * Fixs, update dependencies and copyrights

## 1.3.2 (2021-06-08)
 * Free disk space info, memory limit fixed

## 1.3.1 (2021-05-13)
 * Sphinx Search configuration and support
 * Show modal call fixed

## 1.3.0 (2021-05-04)
 * Public properties for expanding the list of language locales, list of modules available for install and setup, sidebar and creation menu list
 * Animated logotype for Welcome screen
 * Added counter widgets to Dashboard
 * Multiple Sign In`s and logoffs by session timeout
 * Add support for yii2-billing
 * JS fixed for language switcher
 
## 1.2.2 (2021-02-14)
 * Extended dashboard method `getSidebarMenuItems()` to render multi-menu
 * Added support for yii2-catalog and yii2-store modules
 * Added support for yii2-profiles module
 * Added comments widget
 * Menu and routes fixed
 
## 1.2.1 (2020-12-24)
 * Hot keys for pagination
 * Added support for Menu module
 
## 1.2.0 (2020-11-04)
 * Change assets workflow to Gulp.js
 * Refactoring of detect process by grep
 * Added call `phpinfo()` in modal

## 1.1.28 (2020-10-10)
 * Rebuild assets after jQuery update
 * Fixed jQuery version on v3.5.1
 
## 1.1.27 (2020-10-06)
 * Added active processes info. Fixed terminal
 * Update bower dependencies
 * Added support for Guard module
 
## 1.1.26 (2020-09-02)
 * Set error handler, fixed add/activate module in controller
 * Modules controller fixed: add, delete and update module status
 
## 1.1.25 (2020-07-01)
 * Update dependencies, README.md
 
## 1.1.24 (2020-06-06)
 * UrlManager rules fixed
 * Translated description of module (modules list)
 * Added support for Robots.txt module
 * Fontface in dashboard
 * Fixed: Dropdown menu hidding by popover is show
 * Check height of dropdown, dropup variation
 
## 1.1.23 (2020-04-20)
 * Add subclasses for buttons
 * Loads stats widget for dashboard
 
## 1.1.22 (2020-03-25)
 * Added pagination
 * Added DB status to System info
 * Row columns with equal heights
 * Dashboard update, some translations
 * Polyfill for array_key_first() for PHP <= 7.3.0
 
## 1.1.21 (2020-03-06)
 * Stylesheets for dashboard
 * Live search for blog posts in dashboard navbar
 * Added support for Blog module
 
## 1.1.20 (2020-02-27)
 * Added system information page

## 1.1.19 (2020-02-23)
 * Added support for Search module, error handler for dashboard
 
## 1.1.18 (2020-01-20)
 * Added support for RSS, Yandex.Turbo and Google AMP modules
 
## 1.1.17 (2020-01-04)
 * Added shadow in main navbar
 * Load admin assets only for dashboard
 
## 1.1.16 (2019-12-24)
 * Antialiasing and show/hide main navbar/sidebar menu by click
 
## 1.1.15 (2019-12-16)
 * Show/hide main navbar/sidebar menu by hover
 * Storage UI language in cookie
 
## 1.1.14 (2019-12-10)
 * Fixed deprecated class declaration

## 1.1.13 (2019-12-09)
 * Added some module options and fixing password reset email
 
## 1.1.12 (2019-11-30)
 * Added permanently remove module and clear all module options
 
## 1.1.11 (2019-11-26)
 * Added Newsletters, Subscribers and Guard modules support
 
## 1.1.10 (2019-10-28)
 * Added widgets to main dashboard, navigation to main page of site
 
## 1.1.9 (2019-10-15)
 * Add support for translations module
 * Sidebar stylesheets, adaptive nav menu and fix syntax
 
## 1.1.8 (2019-09-09)
 * Fix console
 * Styling sidebar submenu

## 1.1.7 (2019-09-07)
 * Added bug report functionality
 * Fix check updates for modules. Add method getAppVersion() for layout. Update dependencies.
 
## 1.1.6 (2019-09-05)
 * Added module params
 * Check for updates notify for modules
 * Customize navbar menu if some modules not installed
 * Change logotype in navbar

## 1.1.5 (2019-08-31)
 * Some bugfix for migrations and views
 
## 1.1.4 (2019-08-06)
 * Added Mailer and News modules to support

## 1.1.3 (2019-07-29)
 * Added priority of loading modules
 
## 1.1.2 (2019-07-23)
 * Added support of Terminal module 
 
## 1.1.1 (2019-07-20)
 * Loading options from DB for installed modules, fix`s

## 1.1.0 (2019-07-16)
 * Optimize InitController for main installation script
 * Added module manager, migrations

## 1.0.10 (2019-07-11)
 * Styling and managing the sidebar admin menu
 * Sidebar menu functionality
 * Added page loading progress bar
 * Refactoring getSidebarMenuItems() for Dashboard component

## 1.0.9 (2019-07-08)
 * Start stylized sidebar menu
 * Rebuild assets
 
## 1.0.8 (2019-06-10)
 * Added restore and reset password functionality
 
## 1.0.7 (2019-06-08)
 * Module assets and login form
 
## 1.0.6 (2019-06-07)
 * Build assets files and registration for layouts
 * Added admin assets bundle
 * Add NPM package and init Grunt
 
## 1.0.5 (2019-06-07)
 * Module refactoring

## 1.0.4 (2019-06-05)
 * Module transferred to base module interface. Update Yii2 version.

## 1.0.3 (2019-06-02)
 * Bugfix, routing and refactoring

## 1.0.2 (2019-06-01)
 * v.1.1.2 - Update Yii2 version
 
## 1.0.1 (2019-06-01)
 * Added module name and description
 
## 1.0.0 (2019-05-31)
 * Base module, bootstrap

Module.php
<?php

namespace wdmg\admin;

/**
 * Admin dashboard for Butterfly.CMS
 *
 * @category        Module
 * @version         1.4.6
 * @author          Alexsander Vyshnyvetskyy <alex.vyshnyvetskyy@gmail.com>
 * @link            https://github.com/wdmg/yii2-admin
 * @copyright       Copyright (c) 2019 - 2023 W.D.M.Group, Ukraine
 * @license         https://opensource.org/licenses/MIT Massachusetts Institute of Technology (MIT) License
 *
 */

use Yii;
use wdmg\base\BaseModule;
use wdmg\helpers\ArrayHelper;

/**
 * api module definition class
 */
class Module extends BaseModule
{
    /**
     * {@inheritdoc}
     */
    public $controllerNamespace = 'wdmg\admin\controllers';

    /**
     * {@inheritdoc}
     */
    public $defaultRoute = 'admin/index';

    /**
     * @var string, the name of module
     */
    public $name = "Dashboard";

    /**
     * @var string, the description of module
     */
    public $description = "Main administrative panel";

	/**
	 * @var array, the list of support locales for multi-language versions of content.
	 * @note This variable will be override if you use the `wdmg\yii2-translations` module.
	 */
	public $supportLocales = ['ru-RU', 'uk-UA', 'en-US'];

    /**
     * @var boolean, the flag if updates check turn on
     */
    public $checkForUpdates = false;

	public $showDate = true;
    public $showTime = true;
    public $timeFormat24 = true;

    /**
     * @var integer, the time to expire cache
     */
    public $cacheExpire = 3600;
    public $rememberDuration = 86400;
    public $resetTokenExpire = 3600;
    public $supportEmail = 'noreply@example.com';

    /**
     * @var array, expanding the list of language locales for searching translations
     */
    public $customLocales = [];

    /**
     * @var array, expanding the list of modules available for installation and download
     */
    public $customSupportModules = [];

    /**
     * @var array, extending the sidebar menu list
     */
    public $customSidebarMenu = [];

    /**
     * @var array, expanding the creation menu list
     */
    public $customCreateMenu = [];

    /**
     * @var string the module version
     */
    private $version = "1.4.6";

    /**
     * @var integer, priority of initialization
     */
    private $priority = 0;

    /**
     * @var array, support system languages
     */
    private $locales = [
        'en-US' => 'English',
        'ru-RU' => 'Русский',
    ];

    /**
     * @var array of support modules
     */
    private $support = [
        'wdmg/yii2-activity',
        'wdmg/yii2-api',
        'wdmg/yii2-bookmarks',
        'wdmg/yii2-comments',
        'wdmg/yii2-forms',
        'wdmg/yii2-geo',
        'wdmg/yii2-likes',
        'wdmg/yii2-messages',
        'wdmg/yii2-mailer',
        'wdmg/yii2-content',
        'wdmg/yii2-options',
        'wdmg/yii2-catalog',
        'wdmg/yii2-store',
        'wdmg/yii2-billing',
        'wdmg/yii2-menu',
        'wdmg/yii2-pages',
        'wdmg/yii2-news',
        'wdmg/yii2-blog',
        'wdmg/yii2-rbac',
        'wdmg/yii2-reposts',
        'wdmg/yii2-reviews',
        'wdmg/yii2-services',
        'wdmg/yii2-stats',
        'wdmg/yii2-guard',
        'wdmg/yii2-media',
        'wdmg/yii2-redirects',
        'wdmg/yii2-tasks',
        'wdmg/yii2-search',
        'wdmg/yii2-tickets',
        'wdmg/yii2-sitemap',
        'wdmg/yii2-rss',
        'wdmg/yii2-turbo',
        'wdmg/yii2-amp',
        'wdmg/yii2-translations',
        'wdmg/yii2-subscribers',
        'wdmg/yii2-newsletters',
        'wdmg/yii2-terminal',
        'wdmg/yii2-robots',
        'wdmg/yii2-users',
        'wdmg/yii2-profiles',
        'wdmg/yii2-views',
        'wdmg/yii2-votes',
    ];

    /**
     * @var array of sidebar menu items
     */
    private $sidebarMenu = [
        [
            'label' => 'Dashboard',
            'icon' => 'fa fa-fw fa-tachometer-alt',
            'url' => ['/admin/admin/index'],
            'order' => 0,
        ], [
            'label' => 'Modules',
            'icon' => 'fa fa-fw fa-puzzle-piece',
            'url' => ['/admin/admin/modules'],
            'order' => 1,
        ], [
            'label' => 'System',
            'icon' => 'fa fa-fw fa-cogs',
            'items' => [
                'activity',
                'api',
                'options',
                'services',
                'redirects',
                'robots',
                'mailer',
                [
                    'label' => 'Information',
                    'icon' => 'fa fa-fw fa-info-circle',
                    'url' => ['/admin/admin/info'],
                    'order' => 99,
                ],
            ],
            'order' => 2,
        ], /*[
            'label' => 'Access',
            'icon' => 'fa-lock',
            'order' => 3,
            'items' => [
                [
                    'label' => 'Some link3',
                    'icon' => 'fa-check',
                    'items' => ['geo', 'stats'],
                    'order' => 3
                ], [
                    'label' => 'Some link1',
                    'icon' => 'fa-check',
                    'order' => 1
                ], [
                    'label' => 'Some link2',
                    'icon' => 'fa-check',
                    'items' => ['users'],
                    'url' => 'fa-check',
                    'order' => 2
                ]
            ],
        ],*/ [
            'label' => 'Users',
            'icon' => 'fa fa-fw fa-users',
            'items' => ['users', 'profiles', 'rbac'],
            'order' => 14,
        ], [
            'label' => 'Content',
            'icon' => 'fa fa-fw fa-archive',
            'items' => ['pages', 'media', 'content'],
            'order' => 7,
        ], [
            'label' => 'Publications',
            'icon' => 'fa fa-fw fa-pencil-alt',
            'items' => [
                'news',
                'blog',
                'subscribers',
                'newsletters'
            ],
            'order' => 8,
        ], [
            'label' => 'E-commerce',
            'icon' => 'fa fa-fw fa-shopping-bag',
            'items' => ['catalog', 'store', 'billing'],
            'order' => 9,
        ], [
            'label' => 'Feedbacks',
            'icon' => 'fa fa-fw fa-comments',
            'items' => ['reviews', 'comments', 'forms'],
            'order' => 10,
        ], [
            'label' => 'Socials',
            'icon' => 'fa fa-fw fa-share-alt',
            'items' => ['messages', 'likes', 'bookmarks', 'reposts', 'views', 'votes'],
            'order' => 11,
        ], [
            'label' => 'Security',
            'icon' => 'fa fa-fw fa-shield-alt',
            'item' => 'guard',
            'order' => 12,
        ], [
            'label' => 'Common',
            'icon' => 'fa fa-fw fa-wrench',
            'items' => ['menu', 'search', 'geo', 'translations', 'rss', 'turbo', 'amp', 'sitemap'],
            'order' => 13,
        ], [
            'label' => 'Stats',
            'icon' => 'fa fa-fw fa-chart-pie',
            'item' => 'stats',
            'order' => 14,
        ], [
            'label' => 'Support',
            'icon' => 'fa fa-fw fa-life-ring',
            'items' => ['tasks', 'tickets'],
            'order' => 15,
        ]
    ];

    /**
     * @var array of create menu items
     */
    private $createMenu = [
        'wdmg/yii2-pages' => [
            'label' => 'Page',
            'url' => ['/admin/pages/pages/create']
        ], 'wdmg/yii2-media' => [
            'label' => 'Media item',
            'url' => ['/admin/pages/pages/create']
        ], 'wdmg/yii2-content' => [
            [
                'label' => 'Content block',
                'url' => ['/admin/content/blocks/create']
            ],[
                'label' => 'Content list',
                'url' => ['/admin/content/lists/create']
            ],
        ], 'wdmg/yii2-menu' => [
            'label' => 'Menu item',
            'url' => ['/admin/menu/list/create'],
        ], 'wdmg/yii2-news' => [
            'label' => 'News',
            'url' => ['/admin/news/news/create']
        ], 'wdmg/yii2-blog' => [
            'label' => 'Post',
            'url' => ['/admin/blog/posts/create']
        ], 'wdmg/yii2-subscribers' => [
            'label' => 'Subscriber',
            'url' => ['/admin/subscribers/all/create']
        ], 'wdmg/yii2-newsletters' => [
            'label' => 'Newsletter',
            'url' => ['/admin/newsletters/list/create']
        ], 'wdmg/yii2-forms' => [
            'label' => 'Form',
            'url' => ['/admin/forms/list/create']
        ], 'wdmg/yii2-users' => [
            'label' => 'User',
            'url' => ['/admin/users/users/create/']
        ], 'wdmg/yii2-profiles' => [
            'label' => 'Profile',
            'url' => ['/admin/profiles/profiles/create/']
        ], 'wdmg/yii2-tasks' => [
            'label' => 'Task',
            'url' => ['/admin/tasks/item/create/']
        ], 'wdmg/yii2-translations' => [
            'label' => 'Translate',
            'url' => ['/admin/translations/list/create/']
        ],
    ];

    /**
     * @var bool, of show disabled menu items in dashboard
     */
    public $menuShowDisabled = false;

    /**
     * @var bool, of allow to multi Sign In
     */
    public $multiSignIn = false; // not allow by default

    /**
     * @var integer, session timeout in sec. of auth (where `0` is unlimited)
     */
    public $sessionTimeout = 60 * 15; // 15 min.

    /**
     * @var bool, the flag for configuration Sphinx Search
     */
    public $useSphinxSearch = true;

    /**
     * @var array, configuration of Sphinx Search daemon
     */
    public $sphinxSearchConf = [
        'dsn' => "mysql",
        'host' => "127.0.0.1",
        'port' => "9306",
        'username' => "root",
        'password' => "root",
    ];

    /**
     * {@inheritdoc}
     */
    public function init()
    {
        parent::init();

        // Set version of current module
        $this->setVersion($this->version);

        // Set priority of current module
        $this->setPriority($this->priority);

        if (!Yii::$app instanceof \yii\console\Application) {

            // Set authorization route
            Yii::$app->user->loginUrl = ['admin/login'];

            // Set assets bundle, if not loaded
            if ($this->isBackend() && !$this->isConsole()) {

                if (!isset(Yii::$app->assetManager->bundles['wdmg\admin\AdminAsset']))
                    Yii::$app->assetManager->bundles['wdmg\admin\AdminAsset'] = \wdmg\admin\AdminAsset::register(Yii::$app->view);

                if (!isset(Yii::$app->assetManager->bundles['wdmg\admin\FontAwesomeAssets']))
                    Yii::$app->assetManager->bundles['wdmg\admin\FontAwesomeAssets'] = \wdmg\admin\FontAwesomeAssets::register(Yii::$app->view);

            }

            // Check of updates and return current version
            $meta = $this->getMetaData();
            $version = $this->getVersion();
            if ($new_version = $this->checkUpdates($meta['name'], $version)) // Check of updates for `wdmg/yii2-admin`
                $this->view->params['version'] = 'v'. $version . ' <label class="label label-danger">Available update to ' . $new_version . '</label>';
            else
                $this->view->params['version'] = 'v'. $version;

	        $this->showDate = $this->getOption('admin.showDate');
	        $this->showTime = $this->getOption('admin.showTime');
	        $this->timeFormat24 = $this->getOption('admin.timeFormat24');

	        $this->view->params['datetime.showDate'] = $this->showDate;
	        $this->view->params['datetime.showTime'] = $this->showTime;
	        $this->view->params['datetime.timeFormat24'] = $this->timeFormat24;
        }

    }

    /**
     * Return list of support languages
     * @return array of locales
     */
    public function getSupportLanguages()
    {
        return ArrayHelper::merge((is_array($this->customLocales) ? $this->customLocales : []), $this->locales);
    }

    /**
     * Return list of support modules
     * @return array of modules vendor/name
     */
    public function getSupportModules()
    {
        return ArrayHelper::merge((is_array($this->customSupportModules) ? $this->customSupportModules : []), $this->support);
    }

    /**
     * Return list of dashboard main menu items
     * @return array
     */
    public function getMenuItems()
    {
        return ArrayHelper::merge((is_array($this->customSidebarMenu) ? $this->customSidebarMenu : []), $this->sidebarMenu);
    }

    /**
     * Return list of dashboard create menu items
     * @return array
     */
    public function getCreateMenuItems()
    {
        return ArrayHelper::merge((is_array($this->customCreateMenu) ? $this->customCreateMenu : []), $this->createMenu);
    }

    /**
     * Check for available updates
     * @return string, remote version
     */
    public function checkUpdates($module_name, $current_version)
    {
        $viewed = array();
        $session = Yii::$app->session;
        $check_updates = $this->getOption('admin.checkForUpdates');

        if(isset($session['viewed-flash']) && is_array($session['viewed-flash']))
            $viewed = $session['viewed-flash'];

        // Get time to expire cache
        if (isset(Yii::$app->params['admin.cacheExpire']))
            $expire = intval(Yii::$app->params['admin.cacheExpire']);
        else
            $expire = $this->cacheExpire;

        if (!$check_updates) {
            Yii::warning('Attention! In the system settings, the ability to check for updates is disabled.', __METHOD__);
            if(!in_array('admin-check-updates', $viewed) && is_array($viewed)) {
                Yii::$app->getSession()->setFlash(
                    'warning',
                    Yii::t('app/modules/admin', 'Attention! In the system settings, the ability to check for updates is disabled.')
                );
                $session['viewed-flash'] = array_merge(array_unique($viewed), ['admin-check-updates']);
            }
            return false;
        }

        if (Yii::$app->user->isGuest || !$module_name || !$current_version)
            return false;

        $status = null;
        $versions = null;
        $remote_version = null;
        if (Yii::$app->getCache()) {
            if (Yii::$app->cache->exists('modules.versions'))
                $versions = Yii::$app->cache->get('modules.versions');

            if (Yii::$app->cache->exists('modules.updates'))
                $status = Yii::$app->cache->get('modules.updates');
        }
        if (isset($versions[$module_name]))
            $remote_version = $versions[$module_name];

        $update_server = 'https://api.github.com';
        if (is_null($remote_version) && !$status == 'sleep' && @get_headers($update_server)) {

            $client = new \yii\httpclient\Client(['baseUrl' => $update_server]);
            $response = $client->get('/repos/'.$module_name.'/releases/latest', [])->setHeaders([
                'User-Agent' => 'Butterfly.CMS',
                'Content-Type' => 'application/json'
            ])->send();

            if ($response->getStatusCode() == 200) {
                $data = \json_decode($response->content);
                $remote_version = $data->tag_name;

                if ($remote_version && $versions)
                    Yii::$app->cache->add('modules.versions', ArrayHelper::merge($versions, [$module_name => $remote_version]), intval($expire));
                elseif ($remote_version)
                    Yii::$app->cache->add('modules.versions', [$module_name => $remote_version], intval($expire));

            } else {

                if ($response->getStatusCode() == 404) {
                    Yii::error('An error occurred while checking for updates for `'.$module_name.'`. 404 - Resource not found.', __METHOD__);
                    Yii::$app->session->setFlash(
                        'error',
                        Yii::t('app/modules/admin', 'An error occurred while checking for updates for `{module}`. 404 - Resource not found.',
                            ['module' => $module_name]
                        )
                    );
                } else if ($response->getStatusCode() == 403) {

                    if (Yii::$app->getCache())
                        Yii::$app->cache->add('modules.updates', 'sleep', intval($expire));

                    Yii::error('An error occurred while checking for updates to one or more modules. 403 - Request limit exceeded.', __METHOD__);
                    if(!in_array('admin-updates-limit', $viewed) && is_array($viewed)) {
                        Yii::$app->getSession()->setFlash(
                            'error',
                            Yii::t('app/modules/admin', 'An error occurred while checking for updates to one or more modules. 403 - Request limit exceeded.')
                        );
                        $session['viewed-flash'] = array_merge(array_unique($viewed), ['admin-updates-limit']);
                    }
                } else if ($response->getStatusCode() == 503) {
                    if (Yii::$app->getCache())
                        Yii::$app->cache->add('modules.updates', 'sleep', intval($expire));

                    Yii::error('An error occurred while checking for updates to one or more modules. 503 - Service is temporarily unavailable.', __METHOD__);
                    if(!in_array('admin-updates-unavailable', $viewed) && is_array($viewed)) {
                        Yii::$app->getSession()->setFlash(
                            'error',
                            Yii::t('app/modules/admin', 'An error occurred while checking for updates to one or more modules. 503 - Service is temporarily unavailable.')
                        );
                        $session['viewed-flash'] = array_merge(array_unique($viewed), ['admin-updates-unavailable']);
                    }
                }

                return false;
            }
        }

        if (!version_compare($remote_version, $current_version, '<='))
            return $remote_version;
        else
            return false;

    }

}